### PR TITLE
Support Phantom wallet

### PR DESCRIPTION
**Pull request by maximgeerinck**
## Motivation and context

Add support for phantom wallet.

List of current features:

- Connect wallet
- Sign message
- Cancel signing
- Wallet setup
- Copy eth address

EXTRA:

- Multiple providers: Can use metamask and phantom together in 1 session using the PROVIDERS environment variable
- Supports versions using `@`

```
PROVIDERS=phantom,metamask@10.25.0 npx playwright test
```

TODO:
- [x] Phantom to provide public releases (Right now only accessible by providing strongly protected PATs)

## Quality checklist

- [x] I have performed a self-review of my code.

## Future Ideas

- Would be nice if these providers can be classes that extend from an abstract class
- We will look into more generic selectors (like proxying getByText, ... for accessing more generic things in dialogs)
- Typescript support

### CODE REVIEW

**Comment by maximgeerinck** [comment]
> 

We've added a way to download a phantom crx and install it

File: plugins/index.js [hunk 1]
```diff
@@ -117,156 +182,257 @@ module.exports = (on, config) => {
...
-      const accepted = await metamask.acceptAccess(options);
+      return module.exports.acceptAccess(options);
+    },
+    acceptAccess: async options => {
+      const accepted = await getProvider(selectedProvider).acceptAccess(
+        options,
+      );
       return accepted;
     },
     rejectMetamaskAccess: metamask.rejectAccess,
```
[the remark this answers is not in the document]
Comment (maximgeerinck) [reply plugins/index.js]: adjusted

**Comment by maximgeerinck** [comment]
> @maximgeerinck Tests are failing due to the recent changes (mostly adding new networks), can you please take a look? Many thanks for considering my request.

Hey, tried testing, it fails on enabling test networks for metamask but it seems to be passing.
How do you usually run this cause you're skipping setup? if i don't skip setup it seems to work

**Comment by drptbl** [comment]
Hey guys, just wanted to let you know that starting from next month we'll have full time dev working on synpress. I will make sure to prioritise merging this asap. Thanks for understanding. Cheers!

**Comment by Wgil** [comment]
@drptbl those are great news. Do you have any ETA to make synpress wallet agnostic? I would be really interested on using it to test e2e Dapps based on Cosmos chains with wallets like Keplr, Cosmostation or leap wallet

**Comment by drptbl** [comment]
> @drptbl those are great news. Do you have any ETA to make synpress wallet agnostic? I would be really interested on using it to test e2e Dapps based on Cosmos chains with wallets like Keplr, Cosmostation or leap wallet

No ETA, but it should happen within next ~2 months max. We're currently working on adding a plugin system to synpress for wallets (so that each wallet support can be placed in separate node module) and Keplr support is coming after that.

**Comment by safaiyeh** [comment]
GM! Appreciate the initiative @maximgeerinck, we need this at Magic Eden as well. Let's chat? My email is jason@magiceden.io would love to see where we can help.

**Comment by duckception** [comment]
Hey Phantom team! 👋 

I'm the new hire that @drptbl mentioned. 

Let's get this PR merged, shall we? I'll post an initial review at the beginning of next week. In the meantime, could you @maximgeerinck rebase it to the latest changes on our `dev` branch?

File: commands/metamask.js [hunk 2]
```diff
@@ -1332,13 +1490,14 @@ async function activateAdvancedSetting(
       await metamask.goToAdvancedSettings();
     }
   }
-  if ((await playwright.metamaskWindow().locator(toggleOn).count()) === 0) {
-    await playwright.waitAndClick(toggleOff);
+  if (!(await playwright.windows(PROVIDER).locator(toggleOn).count()) === 0) {
```
Comment (duckception) [inline commands/metamask.js:1493]:
```suggestion
  if ((await playwright.windows(PROVIDER).locator(toggleOn).count()) === 0) {
```

File: commands/playwright.js [hunk 3]
```diff
@@ -54,23 +57,108 @@ module.exports = {
...
+            extension.title === 'Phantom Wallet'
+              ? extension.url.replace('popup.html', 'onboarding.html')
+              : extension.url,
+        };
+      })
+      .reduce((prev, curr) => ({ ...prev, [curr.name]: curr }), {});
+
     return browser.isConnected();
   },
+  clearExtensionData: async provider => {
```
Comment (duckception) [inline commands/playwright.js:85]: Any particular reason why this function was introduced?

File: commands/playwright.js [hunk 4]
```diff
@@ -99,44 +187,80 @@ module.exports = {
...
-        metamaskNotificationWindow = page;
+        notificationWindows[provider] = page;
         retries = 0;
         await page.bringToFront();
-        await module.exports.waitUntilStable(page);
-        await module.exports.waitFor(
-          notificationPageElements.notificationAppContent,
-          page,
-        );
+        if (provider == 'metamask') {
```
Comment (duckception) [inline commands/playwright.js:233]:
```suggestion
        if (provider === 'metamask') {
```

File: commands/playwright.js [hunk 5]
```diff
@@ -99,44 +187,80 @@ module.exports = {
...
       throw new Error(
-        '[switchToMetamaskNotification] Max amount of retries to switch to metamask notification window has been reached. It was never found.',
+        `[switchToNotification: ${provider}] Max amount of retries to switch to metamask notification window has been reached. It was never found.`,
       );
     }
   },
-  async waitFor(selector, page = metamaskWindow) {
-    await module.exports.waitUntilStable(page);
+  async waitFor(provider, selector, page = module.exports.windows(provider)) {
+    if (provider == 'metamask') {
```
Comment (duckception) [inline commands/playwright.js:261]:
```suggestion
    if (provider === 'metamask') {
```

File: commands/playwright.js [hunk 6]
```diff
@@ -177,89 +306,146 @@ module.exports = {
...
+  async waitAndSetValue(
+    provider,
+    text,
+    selector,
+    page = module.exports.windows(provider),
+  ) {
+    const element = await module.exports.waitFor(provider, selector, page);
     await element.fill('');
-    await module.exports.waitUntilStable(page);
+    await module.exports.waitUntilStable(provider, page);
```
Comment (duckception) [inline commands/playwright.js:393]: In some functions, you introduce an if statement that runs `.waitUntillStable()` only if the provider is set to MetaMask, yet sometimes you use this function for Phantom too. What's the deal with it?

File: commands/phantom.js [hunk 7]
```diff
@@ -0,0 +1,551 @@
+const log = require('debug')('synpress:phantom');
```
[the remark this answers is not in the document]
Comment (duckception) [reply commands/phantom.js:1]:
```suggestion
```

File: commands/phantom.js [hunk 8]
```diff
@@ -0,0 +1,567 @@
...
+    } else if (chain === 'solana') {
+      await playwright.waitAndClick(
+        PROVIDER,
+        mainPageElements.accountBar.solanaRow,
+      );
+    }
+
+    walletAddress = await playwright
+      .windows(PROVIDER)
+      .evaluate('navigator.clipboard.readText()');
```
Comment (duckception) [inline commands/phantom.js:241]: I'd be very careful with using the raw `clipboard.readText()` function. Let's change this code to "emulate" `CTRL/CMD + V` shortcut.

File: pages/phantom/notification-page.js [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+const gasOptionMediumButton = `${notificationPage} [data-testid="edit-gas-fee-item-medium"]`;
+const gasOptionHighButton = `${notificationPage} [data-testid="edit-gas-fee-item-high"]`;
+const gasOptionDappSuggestedButton = `${notificationPage} [data-testid="edit-gas-fee-item-dappSuggested"]`;
+const gasOptionCustomButton = `${notificationPage} [data-testid="edit-gas-fee-item-custom"]`;
+const baseFeeInput = `${notificationPage} [data-testid="base-fee-input"]`;
+const priorityFeeInput = `${notificationPage} [data-testid="priority-fee-input"]`;
+const editGasLimitButton = `${notificationPage} [data-testid="advanced-gas-fee-edit"]`;
+const gasLimitInput = `${notificationPage} [data-testid="gas-limit-input"]`;
+const saveCustomGasFeeButton = `${notificationPage} .popover-container .btn-primary`;
+const totalLabel = `${confirmPageContent} .transaction-detail-item:nth-child(2) .transaction-detail-item__detail-values h6:nth-child(2)`; // todo: fix
```
Comment (duckception) [inline pages/phantom/notification-page.js:78]: What does this `todo: fix` comment refer to?

File: pages/phantom/main-page.js [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+  container: popupSelector,
+  closeButton: '.popover-header__button',
+  background: '.popover-bg',
+  sendPopup: {
+    container: sendPopupSelector,
+    speedUpButton: `${sendPopupSelector} .btn-primary`,
+    cancelButton: `${sendPopupSelector} .btn-secondary`,
+    transactionStatus: `${sendPopupSelector} .transaction-status`,
+    copyTxIdButton: `${sendPopupSelector} .transaction-list-item-details__tx-hash .transaction-list-item-details__header-button a`,
+    // todo:
```
Comment (duckception) [inline pages/phantom/main-page.js:47]: 👀

File: plugins/index.js [hunk 11]
```diff
@@ -28,15 +42,42 @@ module.exports = (on, config) => {
...
+              '*': {
+                expiration: '0',
+                last_modified: '13248200230459161',
+                model: 0,
+                setting: 1,
+              },
+            },
+          },
+        },
+      };
```
Comment (duckception) [inline plugins/index.js:58]: What is this for?

File: plugins/index.js [hunk 12]
```diff
@@ -28,15 +42,42 @@ module.exports = (on, config) => {
...
     }
 
-    if (!process.env.SKIP_METAMASK_INSTALL) {
-      // NOTE: extensions cannot be loaded in headless Chrome
-      const metamaskPath = await helpers.prepareMetamask(
-        process.env.METAMASK_VERSION || '10.25.0',
-      );
-      arguments_.extensions.push(metamaskPath);
+    if (process.env.PROVIDERS) {
+      process.env.CYPRESS_PROVIDERS = process.env.PROVIDERS;
```
Comment (duckception) [inline plugins/index.js:62]: Why do we set this here?

File: plugins/index.js [hunk 13]
```diff
@@ -336,5 +504,9 @@ module.exports = (on, config) => {
     config.env.SKIP_METAMASK_SETUP = true;
   }
 
+  if (process.env.PROVIDERS) {
+    config.env.PROVIDERS = process.env.PROVIDERS;
+  }
+
```
Comment (duckception) [inline plugins/index.js:510]: Same question. Why do we set it up here?

File: helpers.js [hunk 14]
```diff
@@ -107,7 +108,7 @@ module.exports = {
   },
   getSynpressPath() {
     if (process.env.SYNPRESS_LOCAL_TEST) {
-      return '.';
+      return './node_modules/@phantom/synpress';
```
Comment (duckception) [inline helpers.js:111]:
```suggestion
      return '.';
```

File: plugins/index.js [hunk 15]
```diff
@@ -117,156 +182,259 @@ module.exports = (on, config) => {
...
       return activated;
     },
-    activateImprovedTokenAllowanceInMetamask: async skipSetup => {
-      const activated = await metamask.activateImprovedTokenAllowance(
-        skipSetup,
-      );
+    activateShowCustomNetworkListInMetamask: async skipSetup => {
+      const activated = await getProvider(
+        selectedProvider,
+      ).activateShowCustomNetworkList(skipSetup);
```
Comment (duckception) [inline plugins/index.js:241]: What happened here?

**Review by duckception — CHANGES_REQUESTED** [review]
A little bit late, but here's the initial review 🫡

File: package.json [hunk 16]
```diff
@@ -114,6 +115,9 @@
     "ansi-regex": "5.0.1",
     "@testing-library/dom": "8.20.0"
   },
+  "overrides": {
```
[the remark this answers is not in the document]
Comment (maximgeerinck) [reply package.json:118]: there were some instances were an insecure version of lodash is used, this overrides it

File: commands/playwright.js (hunk 6 above)
Comment (maximgeerinck) [reply commands/playwright.js:393]: i found that for phantom it's not always needed, only when there are animations it's nice to have.

File: pages/phantom/main-page.js (hunk 10 above)
Comment (maximgeerinck) [reply pages/phantom/main-page.js:47]: removed

File: plugins/index.js (hunk 12 above)
Comment (maximgeerinck) [reply plugins/index.js:62]: cause cypress only recognizes env vars that start with CYPRESS_

File: plugins/index.js (hunk 13 above)
Comment (maximgeerinck) [reply plugins/index.js:295]: will remove

File: plugins/index.js (hunk 15 above)
Comment (maximgeerinck) [reply plugins/index.js:189]: nothing much, just using providers now

File: commands/phantom.js (hunk 8 above)
Comment (maximgeerinck) [reply commands/phantom.js:241]: i found this to work rather well though. we're using the browser api so it should be fine, also correct permission was set

File: plugins/index.js (hunk 15 above)
Comment (duckception) [reply plugins/index.js:189]: I mean that `activateImprovedTokenAllowanceInMetamask` function was replaced by `activateShowCustomNetworkListInMetamask`. Curious what happened here

**Comment by duckception** [comment]
Two crucial things:
- MetaMask tests are broken. They need to be fixed. Once this is done, I'll approve workflows on CI.
- Add similar tests for Phantom.

File: plugins/index.js (hunk 15 above)
Comment (maximgeerinck) [reply plugins/index.js:189]: ok, will put that back

**Review by duckception — CHANGES_REQUESTED** [review]
Rebase looks good 👍🏻 I think we only need working tests for both of the wallets and we can merge this 🫡

File: helpers.js [hunk 17]
```diff
@@ -210,37 +234,43 @@ module.exports = {
...
       throw new Error(
-        `[download] Unable to download metamask release from: ${url} to: ${destination} with following error:\n${e}`,
+        `[download] Unable to download provider release from: ${url} to: ${destination} with following error:\n${e}`,
       );
     }
   },
-  async prepareMetamask(version) {
-    const release = await module.exports.getMetamaskReleases(version);
-
+  prepareProvider: async (provider, version) => {
```
Comment (nelitow) [inline helpers.js:241]: Version should not be required as getPhantomReleases doesn't use it?

File: commands/playwright.js [hunk 18]
```diff
@@ -149,57 +203,85 @@ module.exports = {
...
     await sleep(200);
     if (retries < 50) {
       retries++;
-      return await module.exports.switchToMetamaskNotification();
+      return await module.exports.switchToNotification(provider);
     } else if (retries >= 50) {
       retries = 0;
       throw new Error(
-        '[switchToMetamaskNotification] Max amount of retries to switch to metamask notification window has been reached. It was never found.',
+        `[switchToNotification: ${provider}] Max amount of retries to switch to metamask notification window has been reached. It was never found.`,
```
Comment (nelitow) [inline commands/playwright.js:277]: Suggestion: Max amount of retries to switch to ${provider} {...}

**Comment by maximgeerinck** [comment]
Closing due to inactivity

File: commands/phantom.js [hunk 19]
```diff
@@ -0,0 +1,645 @@
...
+      notificationPage,
+      { waitForEvent: 'close' },
+    );
+    return true;
+  },
+  confirmTransaction: async () => {
+    const notificationPage = await playwright.switchToNotification(PROVIDER);
+    await playwright.waitAndClick(
+      PROVIDER,
+      transactionPageElements.buttons.rejectTransaction,
```
Comment (nelitow) [inline commands/phantom.js:402]: confirm is rejecting the transaction

**Pull request closed without merging**